src/Attribute/EarlyValidation.php

```diff
@@ -0,0 +1,38 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Input\Validation\Attribute;
+
+use Attribute;
+use Yiisoft\Hydrator\ParameterAttributeInterface;
+use Yiisoft\Validator\RuleInterface;
+
+#[Attribute(Attribute::TARGET_PROPERTY | Attribute::TARGET_PARAMETER | Attribute::IS_REPEATABLE)]
+final class EarlyValidation implements ParameterAttributeInterface
```

Comment:
```suggestion
final class EarlyValidation implements ParameterAttributeInterface, RulesProviderInterface
```
?

Reply:
Why? It's attribute that never will used as implementation of `RulesProviderInterface`.

Reply:
Decided not to do this.